EventFiltering/PWGCF/CFFilter.cxx

```diff
@@ -0,0 +1,178 @@
+///
+/// \author Laura Serksnyte, TU München, laura.serksnyte@cern.ch
+
+#include "Framework/runDataProcessing.h"
+#include "Framework/AnalysisDataModel.h"
+#include "Framework/AnalysisTask.h"
+#include "Framework/ASoAHelpers.h"
+#include "Framework/HistogramRegistry.h"
+
+#include "../filterTables.h"
+
+#include "../../PWGCF/DataModel/FemtoDerived.h"
+#include "../../PWGCF/FemtoDream/FemtoDreamParticleHisto.h"
+#include "../../PWGCF/FemtoDream/FemtoDreamPairCleaner.h"
+#include "../../PWGCF/FemtoDream/FemtoDreamContainer.h"
+#include "../../PWGCF/FemtoDream/FemtoDreamMath.h"
+#include "../../PWGCF/FemtoDream/FemtoDreamPairCleaner.h"
+
+#include "Common/DataModel/EventSelection.h"
+#include "Common/DataModel/Multiplicity.h"
+
+#include <cmath>
+#include <string>
+
+namespace
+{
+
+static constexpr int nTriplets{4};
+
+enum CFTriggers {
+  kPPP = 0,
+  kPPL,
+  kPLL,
+  kLLL
+};
+
+static const std::vector<std::string> CfTriggerNames{"ppp", "ppL", "pLL", "LLL"};
+
+} // namespace
+
+namespace o2::aod
+{
+using FullCollision = soa::Join<aod::Collisions,
+                                aod::EvSels,
+                                aod::Mults>::iterator;
+} // namespace o2::aod
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+using namespace o2::analysis::femtoDream;
+
+struct CFFilter {
+
+  Produces<aod::CFFilters> tags;
+
+  HistogramRegistry registry{"registry", {}, OutputObjHandlingPolicy::AnalysisObject, true, true};
+
+  //FemtoDreamPairCleaner<aod::femtodreamparticle::ParticleType::kTrack, aod::femtodreamparticle::ParticleType::kTrack> pairCleanerTT; Currently not used, will be needed later
+  FemtoDreamPairCleaner<aod::femtodreamparticle::ParticleType::kTrack, aod::femtodreamparticle::ParticleType::kV0> pairCleanerTV;
+
+  void init(o2::framework::InitContext&)
+  {
+    registry.add("fProcessedEvents", "CF - event filtered;;events", HistType::kTH1F, {{6, -0.5, 4.5}});
+    std::array<std::string, 6> eventTitles = {"all", "rejected", "p-p-p", "p-p-L", "p-L-L", "L-L-L"};
+    for (size_t iBin = 0; iBin < eventTitles.size(); iBin++) {
+      registry.get<TH1>(HIST("fProcessedEvents"))->GetXaxis()->SetBinLabel(iBin + 1, eventTitles[iBin].data());
+    }
+    registry.add("fSameEvent", "CF - same event distribution;;events", HistType::kTH1F, {{8000, 0, 8}});
+  }
+
+  float mMassProton = TDatabasePDG::Instance()->GetParticle(2212)->Mass();
+  float mMassLambda = TDatabasePDG::Instance()->GetParticle(3122)->Mass();
+
+  void process(aod::FullCollision const& col, o2::aod::FemtoDreamCollisions const& colsFemto,
+               o2::aod::FemtoDreamParticles& partsFemto)
+  {
+    registry.get<TH1>(HIST("fProcessedEvents"))->Fill(0);
+    bool keepEvent[nTriplets]{false};
+
+    // check if current collision was selected by femto skimmer
+    auto collsInFemto = colsFemto.sliceBy(o2::aod::femtodreamcollision::collisionId, col.globalIndex());
```

Comment:
You are right that this is inefficient. We use a slice to find a single value.
First of all there is no parallel processing on this level. A process function will always process a full data frame.

I would add a setting `fillAll` to the femto framework that changes how `FemtoDreamCollisions` is produced. In this case `FemtoDreamCollisions` is filled for *all* collisions and it can be therefore joined with `FullCollision`. If the collision is not interesting for femto then `FemtoDreamParticles` is not filled for that collision.

Then you subscribe here to `process(join<FemtoDreamCollisions, FullCollisions>::iterator, FemtoDreamParticles`.

You profit from the full grouping before process is called. If FemtoDreamParticles is empty, you make the event as not triggered, otherwise you do the processing as you do it now.

Let me know what you think